api/test/e2e/journeys/test_schemas.py

```diff
@@ -52,23 +45,116 @@ def update_schema(self, schema: dict) -> Response:
+        assert sorted(column_names) == sorted(rapid_schema_column_names_v1)
 
         schema_v2 = self.read_schema_version(SchemaVersion.V2)
         response2 = self.update_schema(schema_v2)
         assert response2.status_code == HTTPStatus.OK
 
-        # TODO: Some kind of assert that the new schema is the second one
+        dataset_schema_v2 = self.get_dataset_from_dynamodb(layer, domain, dataset)
+        rapid_schema_column_names_v2 = self.extract_column_names(dataset_schema_v2)
+
+        assert dataset_schema_v2[0]["domain"] == schema_v2["metadata"]["domain"]
+        assert dataset_schema_v2[0]["layer"] == schema_v2["metadata"]["layer"]
+        assert sorted(column_names) == sorted(rapid_schema_column_names_v2)
+
+    def test_invalid_schema_upload(self):
+        invalid_metadata = {
+            "layer": "default123",
+            "sensitivity": "INVALID",
+            "owners": [{"email": "change_me@email.com"}],
+            "update_behaviour": "INVALID",
+        }
+
+        invalid_metadata_status_codes = []
+        for k, v in invalid_metadata.items():
+            schema_v1 = self.read_schema_version(SchemaVersion.V1)
+            schema_v1["metadata"]["dataset"] = "invalid_dataset"
+            schema_v1["metadata"][k] = v
+            response = self.upload_schema(schema_v1)
+            invalid_metadata_status_codes.append(response.status_code)
+
+        assert all(
+            status_code == HTTPStatus.BAD_REQUEST
+            for status_code in invalid_metadata_status_codes
+        )
+
+        invalid_columns = {"data_type": "INVALID", "allow_null": "pretend_boolean"}
+
+        for k, v in invalid_columns.items():
+            schema_v1 = self.read_schema_version(SchemaVersion.V1)
+            schema_v1["metadata"]["dataset"] = "invalid_dataset"
+            schema_v1["columns"][0][k] = v
+            response = self.upload_schema(schema_v1)
+            assert response.status_code == HTTPStatus.BAD_REQUEST
+
+        schema_v1 = self.read_schema_version(SchemaVersion.V1)
+        schema_v1.pop("columns", None)
+        response1 = self.upload_schema(schema_v1)
+        assert response1.status_code == HTTPStatus.BAD_REQUEST
+
+    def test_invalid_schema_update(self):
+        schema_v2 = self.read_schema_version(SchemaVersion.V2)
+        schema_v2.pop("columns", None)
+        response2 = self.update_schema(schema_v2)
+        assert response2.status_code == HTTPStatus.BAD_REQUEST
+
+    def test_generates_new_schema(self):
+        files = {
+            "file": (self.csv_filename, open("./test/e2e/" + self.csv_filename, "rb"))
+        }
+        generate_url = self.schema_generate_url(
+            self.layer, "PUBLIC", self.e2e_test_domain, "test"
+        )
+
+        response = requests.post(
+            generate_url, headers=self.generate_auth_headers(), files=files
+        )
+
+        assert response.status_code == HTTPStatus.OK
```

Comment:
I think we should also assert that the schema is what we expect

Reply:
Done. This isn't super trivial, the generate schema endpoint creates schemas with some placeholders etc -> have created a new schema.json.tpl file for this.